src/macis/gf/bandlan.cxx

```diff
@@ -196,88 +120,30 @@ bool GetEigsys(std::vector<std::vector<double> > &mat,
   eigvals.clear();
   eigvecs.clear();
   // PREPARE VARIABLES FOR LAPACK
-  char UPLO = 'U', COMPZ = 'V';
-  int N = mat.size(), LDA = mat.size(), LWORK = -1, INFO;
-  double *A, *D, *E, *TAU, *WORK;
+  lapack::Uplo UPLO = lapack::Uplo::Upper;
+  lapack::Job JOBZ = lapack::Job::Vec;
+  int N = mat.size(), LDA = mat.size();
+  std::vector<double> A, D;
 
   // INITIALIZE A
-  A = new double[N * N];
+  A.resize(N * N);
   for(int i = 0; i < N; i++) {
     for(int j = 0; j < N; j++) A[i + j * N] = mat[i][j];
     mat[i].clear();
   }
   mat.clear();
   // ALLOCATE REST OF THE MEMORY
-  D = new double[N];
-  E = new double[N - 1];
-  TAU = new double[N - 1];
-  WORK = new double[N];
-
-  // TRANSFORM THE MATRIX TO TRIDIAGONAL FORM
-  // FIRST, PERFORM MEMORY QUERY
-  dsytrd_(&UPLO, &N, A, &LDA, D, E, TAU, WORK, &LWORK, &INFO);
-  if(INFO != 0) {
-    std::cout << "ERROR IN dsytrd_ MEMORY QUERY!! ERROR CODE: " << INFO
-              << std::endl;
-    return false;
-  }
-  LWORK = WORK[0];
-  delete[] WORK;
-  WORK = new double[LWORK];
-  // NOW, TRANSFORM MATRIX TO TRIDIAGONAL FORM
-  dsytrd_(&UPLO, &N, A, &LDA, D, E, TAU, WORK, &LWORK, &INFO);
-  if(INFO != 0) {
-    std::cout
-        << "ERROR IN dsytrd_ COMPUTING THE TRIDIAGONAL MATRIX!! ERROR CODE: "
-        << INFO << std::endl;
-    return false;
-  }
-
-  // COMPUTE THE TRANSFORMATION MATRIX, NECESSARY TO COMPUTE EIGENVECTORS
-  // FIRST, PERFORM MEMORY QUERY
-  LWORK = -1;
-  dorgtr_(&UPLO, &N, A, &LDA, TAU, WORK, &LWORK, &INFO);
-  if(INFO != 0) {
-    std::cout << "ERROR IN dorgtr_ MEMORY QUERY!! ERROR CODE: " << INFO
-              << std::endl;
-    return false;
-  }
-  LWORK = WORK[0];
-  delete[] WORK;
-  WORK = new double[LWORK];
-  // NOW, COMPUTE THE TRANSFORMATION MATRIX. IT WILL BE STORED IN A
-  dorgtr_(&UPLO, &N, A, &LDA, TAU, WORK, &LWORK, &INFO);
-  if(INFO != 0) {
-    std::cout
-        << "ERROR IN dorgtr_ COMPUTING TRANSFORMATION MATRIX!! ERROR CODE: "
-        << INFO << std::endl;
-    return false;
-  }
-  delete[] TAU;
-
-  // FINALLY, COMPUTE THE EIGENVALUES AND EIGENVECTORS!
-  delete[] WORK;
-  WORK = new double[2 * N - 2];
-  dsteqr_(&COMPZ, &N, D, E, A, &LDA, WORK, &INFO);
-  if(INFO != 0) {
-    std::cout << "ERROR IN dsteqr_ COMPUTING EIGENVECTORS AND EIGENVALUES!! "
-                 "ERROR CODE: "
-              << INFO << std::endl;
-    return false;
-  }
-  delete[] WORK;
-  delete[] E;
+  D.resize(N);
+  lapack::heev_2stage(JOBZ, UPLO, N, A.data(), LDA, D.data());
```

Comment:
Good call on the `2stage`, but that' actually isn't faster for all general cases. Brings up an interesting point about how to select the algorithm though... Might be good to have a switch. I'll give it some thought.